openid-4-verifiable-credential-issuance-1_0.md

```diff
@@ -314,7 +314,9 @@ For security considerations, see (#credential-offer-security).
 This specification defines the following parameters for the Credential Offer object:
 
 * `credential_issuer`: REQUIRED. The URL of the Credential Issuer, as defined in (#credential-issuer-identifier), from which the Wallet is requested to obtain one or more Credentials. The Wallet uses it to obtain the Credential Issuer's Metadata following the steps defined in (#credential-issuer-wellknown).
-* `credentials`: REQUIRED. A JSON array, where every entry is a JSON object or a JSON string. If the entry is an object, the object contains the data related to a certain credential type the Wallet MAY request. Each object MUST contain a `format` Claim determining the format of the credential to be requested and further parameters characterizing the type of the credential to be requested as defined in (#format_profiles). If the entry is a string, the string value MUST be one of the `scope` values in one of the objects in the `credentials_supported` Credential Issuer metadata parameter. When processing, the Wallet MUST resolve this string value to the respective object.
+* `credentials`: REQUIRED. A JSON array, where every entry is a JSON string or a JSON object describing a credential the Wallet MAY request. If the entry is a string, the string value MUST be one of the `scope` values included in a `credentials_supported` Credential Issuer metadata parameter as defined in (#credential-metadata-object). When processing, the Wallet MUST resolve this string value to the respective object. If the entry is an object, the object contains the data related to a certain credential type. Each object contains the following parameters: 
+  * `format`: REQUIRED. JSON string determining the format of the credential.
+  * Parameters characterizing the type of the credential to be requested: REQUIRED. These parameters are specific to the credential format profile, some of which are defined in (#format_profiles).
```

Comment:
I wonder if an identifier should also be allowed in the offer credentials object, as a way to precisely specify the credential to issue.
Use case:
- Training company has a web site where users can see the completed training courses.
- For each course (e.g. `AERO-123-FluidMechanics-II-2023`), user can require a credential, which requires some sort of payment.
- After payment transaction is completed, web site presents an offer as a QR code. This offer exactly identifies the credentials that the user can request (e.g. `AERO-123-FluidMechanics-II-2023`).
- User scans QR code with wallet, authenticates, consents on a screen that only shows `AERO-123-FluidMechanics-II-2023` credential information (no need for the user to select anything), and finally gets an issued credential.

The use of an identifier would allow an offer to be much more precise in the credential that is being offered.

The identifier could be carried in the authorization request's `authorization_details` object.

Reply:
That proposal would require to send the identifiers with the authorization and/or token request, too. It looks complex on first sight but could further simplify the live of implementers as the wallet could _just_ use the identifiers to request the authz code or the access token. This requires us to redefine the authorization details type to allow to carry identifier only (very much like what I proposed for the credential endpoint). I'm not sure how it would work with scopes, I guess we would need to allow authorization request with a list of identifiers w/o scope parameter at all. 
Note: I think it makes sense for pre authz code and authorization code with issuer state, I don't see how it could be implemented with authz code w/o issuer state. 
PS: I would trade that feature for dropping scopes at all ;-)

Reply:
I agree it is beneficial to add identifiers in the credential offer. it will also allow the wallet to look up a specific credential's metadata in the issuer metadata. 

@tlodderstedt when scopes are used, I would expect the following:
- wallet uses an identifier to look up metadata about an offered credential. not scope because there is no guarantee scope value is unique to a credential. (there is a related comment from Pedro below)
- wallet uses the scope value discovered in the previous step in authz/token requests.
- the rest is as already defined.

when authorization details is used, identifier should probably be included inside the `authorization_details` parameter passed in authz/token requests (token req would be optional)?

Reply:
I think no need for a new issue. discussion is happening in issue #77